### PR TITLE
remove die calls that can affect single processes or radiusd

### PR DESCRIPTION
**Pull request by julsemaan**
# Description

Removes calls to die when running in a single process like pfdns or when running in radiusd extension point.
Here is the commented grep report : http://pastebin.com/cnQKv55V 
Feel free to look at it and review the decisions
# Impacts

Lots of stuff
# Delete branch after merge

YES
# NEWS file entries
## Enhancements
- Reworked libraries to remove calls to die as it can kill processes or FreeRADIUS threads

### CODE REVIEW

**Comment by jrouzierinverse** [comment]
What about when a third party library dies?

It would be a lot easier  to wrap our run loops in an eval then try to get rid of all the places where we die in our code.

**Comment by julsemaan** [comment]
We can add this also but also taking care of not over-using die is also important.

**Comment by jrouzierinverse** [comment]
If the goal is to protect the run loops in long running processes then this should be done first.
Then we can start looking on the proper use of dying in our code.

**Comment by julsemaan** [comment]
I agree with that. 

We found it to be a good idea to also rethink some of the dying so this PR comes from there

File: lib/pf/factory/config.pm [hunk 1]
```diff
@@ -39,7 +40,8 @@ sub new {
         return $object;
     }
     else {
-        die "$type is not a valid type";
+        get_logger->error("$type is not a valid type");
+        return undef;
```
Comment (jrouzierinverse) [inline lib/pf/factory/config.pm:44]: Is this module even used?

File: lib/pf/factory/firewallsso.pm [hunk 2]
```diff
@@ -38,12 +39,18 @@ sub new {
...
+    unless(defined $type) {
+        get_logger->error("Type is not defined");
+        return undef;
+    }
     my $subclass = "${mainClass}::${type}";
-    die "type is not defined for $name" unless defined $type;
-    die "$type is not a valid type" unless any { $_ eq $subclass  } @MODULES;
+    unless (any {$_ eq $subclass} @MODULES){
+        get_logger->error("Type $type is not valid");
+        return undef;
```
Comment (jrouzierinverse) [inline lib/pf/factory/firewallsso.pm:52]: Add additional error handling to where getModuleName is called.

File: lib/pf/factory/pki_provider.pm [hunk 3]
```diff
@@ -52,12 +53,18 @@ Get the sub module pf::pki_provider base off it's configuration
...
+    }
     my $subclass = "${mainClass}::${type}";
-    die "type is not defined for $name" unless defined $type;
-    die "$type is not a valid type" unless any { $_ eq $subclass  } @MODULES;
+    unless (any {$_ eq $subclass} @MODULES){
+        get_logger->error("Type $type is not valid");
+        return undef;
+    }
     $subclass;
 }
```
Comment (jrouzierinverse) [inline lib/pf/factory/pki_provider.pm:69]: Add additional error handling to where getModuleName is called.

File: lib/pf/factory/profile/filter.pm [hunk 4]
```diff
@@ -52,9 +53,15 @@ sub getModuleName {
...
+        return undef;
+    }
     my $subclass = "${mainClass}::${type}";
-    die "$type is not a valid type" unless any {$_ eq $subclass} @MODULES;
+    unless (any {$_ eq $subclass} @MODULES){
+        get_logger->error("Type $type is not valid");
+        return undef;
+    }
     $subclass;
 }
```
Comment (jrouzierinverse) [inline lib/pf/factory/profile/filter.pm:66]: Add additional error handling to where getModuleName is called.

File: lib/pf/factory/provisioner.pm [hunk 5]
```diff
@@ -36,12 +37,18 @@ sub new {
...
+    }
     my $subclass = "${mainClass}::${type}";
-    die "type is not defined for $name" unless defined $type;
-    die "$type is not a valid type" unless any { $_ eq $subclass  } @MODULES;
+    unless (any {$_ eq $subclass} @MODULES){
+        get_logger->error("Type $type is not valid");
+        return undef;
+    }
     $subclass;
 }
```
Comment (jrouzierinverse) [inline lib/pf/factory/provisioner.pm:53]: Add additional error handling to where getModuleName is called.

File: lib/pf/factory/scan.pm [hunk 6]
```diff
@@ -38,12 +39,18 @@ sub new {
...
+    }
     my $subclass = "${mainClass}::${type}";
-    die "type is not defined for $name" unless defined $type;
-    die "$type is not a valid type" unless any { $_ eq $subclass  } @MODULES;
+    unless (any {$_ eq $subclass} @MODULES){
+        get_logger->error("Type $type is not valid");
+        return undef;
+    }
     $subclass;
 }
```
Comment (jrouzierinverse) [inline lib/pf/factory/scan.pm:55]: Add additional error handling to where getModuleName is called.

File: lib/pf/factory/triggerParser.pm [hunk 7]
```diff
@@ -32,11 +33,19 @@ sub new {
...
+        return undef;
+    }
     my $subclass = "${mainClass}::${type}";
-    die "Invalid trigger type $type" unless any { $_ eq $subclass  } @MODULES;
-    return $subclass;
+    unless (any {$_ eq $subclass} @MODULES){
+        get_logger->error("Type $type is not valid");
+        return undef;
+    }
+    $subclass;
```
Comment (jrouzierinverse) [inline lib/pf/factory/triggerParser.pm:48]: Add additional error handling to where getModuleName is called.

File: lib/pf/inline/accounting.pm [hunk 8]
```diff
@@ -211,7 +214,10 @@ sub inline_accounting_maintenance {
         # Commit database transaction
         unless ($dbh->commit) {
             $logger->error("Error while committing database transaction: " . $dbh->errstr);
-            $dbh->rollback or $logger->logdie("Can't rollback database transaction: " . $dbh->errstr);
+            unless($dbh->rollback){
+                $logger->error("Can't rollback database transaction: " . $dbh->errstr);
+                return undef;
+            }
         }
     }
```
Comment (jrouzierinverse) [inline lib/pf/inline/accounting.pm:222]: Check to see if additional error handling is needed where inline_accounting_maintenance is called

File: lib/pf/radius/rpc.pm [hunk 9]
```diff
@@ -45,11 +45,10 @@ sub send_rpc_request {
...
             $response = Data::MessagePack->unpack($response_body);
         } else {
-            die "An error occured while processing the MessagePack request return code ($response_code)";
+            return undef;
         }
     } else {
-        my $msg = "An error occured while sending a MessagePack request: $curl_return_code ".$curl->strerror($curl_return_code)." ".$curl->errbuf;
-        die $msg;
+        return undef;
     }
```
Comment (jrouzierinverse) [inline lib/pf/radius/rpc.pm:52]: If there is an error sending an rpc message it will no longer be logged.

File: lib/pf/soh.pm [hunk 10]
```diff
@@ -571,8 +577,15 @@ sub _identify_os {
...
+        return undef;
+    }
+    unless($major){
+        $self->{logger}->error("No operating system version specified");
+        return undef;
+    }
+
 
     my $os;
     unless ($vendor eq 'Microsoft') {
```
Comment (jrouzierinverse) [inline lib/pf/soh.pm:591]: Add additional error handling to where _identify_os is called

File: lib/pf/soh.pm [hunk 11]
```diff
@@ -258,7 +263,8 @@ sub parse_request {
         }
 
         unless (%ok) {
-            die "Couldn't parse any SoH-MS-Windows-Health-Status lines";
+            $self->{logger}->error("Couldn't parse any SoH-MS-Windows-Health-Status lines");
+            return undef;
```
Comment (jrouzierinverse) [inline lib/pf/soh.pm:267]: This is wrapped in try {} catch {} block returning in the try block does not exit the function.

File: lib/pf/util/dhcp.pm [hunk 12]
```diff
@@ -301,7 +301,10 @@ sub make_pcap_filter {
     #listen to all if no types are provided
     return "udp and (port 67 or port 68 or port 767)" unless @types;
     for my $type (@types) {
-       die "Unknown message type $type" unless exists $MESSAGE_TYPE{$type} && defined $MESSAGE_TYPE{$type};
+        unless (exists $MESSAGE_TYPE{$type} && defined $MESSAGE_TYPE{$type}) {
+            get_logger->error("Unknown message type $type" );
+            return "udp and (port 67 or port 68 or port 767)";
```
Comment (jrouzierinverse) [inline lib/pf/util/dhcp.pm:306]: "udp and (port 67 or port 68 or port 767)" as a constant?

File: lib/pf/version.pm [hunk 13]
```diff
@@ -114,9 +114,10 @@ i.e: PacketFence X.Y.Z
...
-    open( $pfrelease_fh, '<', "$conf_dir/pf-release" )
-        || get_logger->logdie("Unable to open $conf_dir/pf-release: $!");
-
+    unless (open( $pfrelease_fh, '<', "$conf_dir/pf-release" )){
+        get_logger->error("Unable to open $conf_dir/pf-release: $!");
+        return undef;
+    }
     $release = <$pfrelease_fh>;
     close($pfrelease_fh);
     chomp($release);
```
Comment (jrouzierinverse) [inline lib/pf/version.pm:123]: Add additional error handling to where version_get_release is called

**Comment by louismunro** [comment]
Rebase?

**Comment by extrafu** [comment]
Get some traction on this PR so we can merge it and be ready for v6.

**Comment by jrouzierinverse** [comment]
In my opinion it should be closed or completely reworked.

**Comment by julsemaan** [comment]
I think closed like you

**Pull request closed without merging**